mayastor/src/subsys/mbus/registration.rs

```diff
@@ -0,0 +1,203 @@
+        let (msg_sender, msg_receiver) = smol::channel::unbounded::<()>();
+        let config = Configuration {
+            node: node.to_owned(),
+            grpc_endpoint: grpc_endpoint.to_owned(),
+            hb_interval: match env::var("MAYASTOR_HB_INTERVAL")
+                .map(|v| v.parse::<u64>())
+            {
+                Ok(Ok(num)) => Duration::from_secs(num),
+                _ => HB_INTERVAL,
+            },
+        };
+        Self {
+            config,
+            rcv_chan: msg_receiver,
+            fini_chan: msg_sender,
+        }
+    }
+}
+
+impl Registration {
+    /// Connect to the server and start emitting periodic register
+    /// messages.
+    /// Runs until the sender side of the message channel is closed
+    pub async fn run_loop(&mut self) {
+        info!(
+            "Registering '{}' and grpc server {} ...",
+            self.config.node, self.config.grpc_endpoint
+        );
+        loop {
+            if let Err(err) = self.register().await {
+                error!("Registration failed: {:?}", err);
+            };
+
+            select! {
+                _ = smol::Timer::after(self.config.hb_interval).fuse() => continue,
+                msg = self.rcv_chan.next().fuse() => {
+                    match msg {
+                        Some(_) => log::info!("Messages have not been implemented yet"),
+                        _ => {
+                            log::info!("Terminating the registration handler");
+                            break;
+                        }
+                    }
+                }
+            };
+        }
+        if let Err(err) = self.deregister().await {
+            error!("Deregistration failed: {:?}", err);
+        };
+    }
+
+    /// Send a register message to the MessageBus.
+    async fn register(&self) -> Result<(), Error> {
+        let payload = RegisterArgs {
+            id: self.config.node.clone(),
+            grpc_endpoint: self.config.grpc_endpoint.clone(),
+        };
+        super::message_bus()
+            .publish("register", serde_json::to_vec(&payload).unwrap())
+            .await
+            .map_err(|cause| Error::QueueRegister {
+                cause,
+            })?;
+
+        // Note that the message was only queued and we don't know if it was
+        // really sent to the message server
+        // We could explicitly flush to make sure it reaches the server or
+        // use request/reply to guarantee that it was delivered
+        debug!(
+            "Registered '{}' and grpc server {}",
+            self.config.node, self.config.grpc_endpoint
+        );
+        Ok(())
+    }
+
+    /// Send a deregister message to the MessageBus.
+    async fn deregister(&self) -> Result<(), Error> {
+        let payload = DeregisterArgs {
+            id: self.config.node.clone(),
+        };
+        super::message_bus()
+            .publish("deregister", serde_json::to_vec(&payload).unwrap())
```

Comment:
Same as above, we can make the argument have a trait guard so we can do the json calls inside the publish function